### PR TITLE
Adding Toasts

### LINKED ISSUE #3530: Add toast component in notifications
### Problem to solve

Notifications specs in https://github.com/gitpod-io/gitpod/issues/3398 included a toast component for visual feedback when a user changes their notification settings but wasn't included in https://github.com/gitpod-io/gitpod/pull/3513.

The toast component will be also useful in other pages like Integrations, see https://github.com/gitpod-io/gitpod/issues/3401.

### User Experience goal

The user should be able to change notifications settings while having visual feedback upon change.

See also [Don’t Prioritize Efficiency Over Expectations](https://www.nngroup.com/articles/efficiency-vs-expectations/):

> Users need to be continuously aware of the current status of the system. For users, seeing is believing: visual feedback must be displayed for every process a system undertakes. 

### Proposal

Let's a) implement a toast component and b) add it on notifications page.

| Notifications |
|-|
| image |

### PR DESCRIPTION
**Pull request by selfcontained**
## Description

**WIP**
Adding Toasts and updating User Preferences to use them for dotfiles and timeout fields. Also tweaked those forms on user prefs page to have a bit more feedback for the UX.

Some considerations:
* Opted for animate in from right for toasts as it's a slightly simpler thing to implement that centered toasts that animate in from bottom. We can always revisit though.
* Deferred adding specific types of toasts to reduce initial scope, i.e. success/error/info etc. We should add those as a followup though.

| 1 | 2 |
|--------|--------|
| image | image |
| image | image | 

## Related Issue(s)

Fixes WEB-38
Fixes #3530

## How to test

* Testing updating your dotfiles repo and default workspace timeout

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by geropl** [comment]
Happened to bump into this: Seems we're still using the an `alert` for the error case:
image

Not sure if this is in-scope for this PR, but: Would it make sense to make that symmetric? Or is there a better way for error messages, e.g. closer to the input field? :thinking:

**Comment by gtsiolis** [comment]
Looking at this now! 👀

**Review by gtsiolis — APPROVED** [review]
Excited to see the toast notifications coming alive, @selfcontained! 🌮 🌮 

Left some UX comments below, but feel free to open follow-up issues or combine topics in a follow-up issue or PR. 🏓  

Approving to unblock merging, but holding in case you'd like to address any of the comments in this PR.

/hold

**Comment by selfcontained** [comment]
> Happened to bump into this: Seems we're still using the an `alert` for the error case: image
> 
> Not sure if this is in-scope for this PR, but: Would it make sense to make that symmetric? Or is there a better way for error messages, e.g. closer to the input field? 🤔

My plan was to make that an error styled toast, but in a follow up where I can work with @gtsiolis around what those styled toasts might look like. It would be easy to make that a toast instead in this PR too though.

**Comment by gtsiolis** [comment]
Regarding the browser alert[[1](https://github.com/gitpod-io/gitpod/pull/17030#issuecomment-1488258114)][[2](https://github.com/gitpod-io/gitpod/pull/17030#issuecomment-1488967092)], this was mentioned also in the original PR in https://github.com/gitpod-io/gitpod/pull/16503#discussion_r1117333268 and this is currently the UX on production. It should be perfectly fine to skip fixing this in the scope of this PR. 💭

**Review by geropl — APPROVED** [review]
Thanks for the explanations. :pray: 

Code LGTM, tested and works. :lotus_position:

File: components/dashboard/tailwind.config.js
Comment (selfcontained) [reply components/dashboard/tailwind.config.js:59]: Yah we can. fwiw I tried coming in from the bottom and centered, and it looked a little janky w/ stacked toasts and how they would shift up to make room instantly while the new one slides up. We could make it pretty slick, but will be a bit more effort to animate it nicely I think. In from right looks a little better with this lower level of effort imo, so I ended up doing that. It's not too tough to swap it to in from bottom to show you what I mean sometime too if you wanted to see each.

File: components/dashboard/src/user-settings/Preferences.tsx
Comment (selfcontained) [reply components/dashboard/src/user-settings/Preferences.tsx:90]: Good eye 🦅 👁️  - it was a side effect of re-using the form components here, and finding a way to add the button into the row the input was on, but keep the `hint` rendering as it did by default. I'll play around with this a little more to see if there's a better way to have that behavior, but still re-use the form input components.

File: components/dashboard/src/components/toasts/Toast.tsx
Comment (selfcontained) [reply components/dashboard/src/components/toasts/Toast.tsx:16]: Nice idea, definitely!

File: components/dashboard/src/components/toasts/Toasts.tsx
Comment (selfcontained) [reply components/dashboard/src/components/toasts/Toasts.tsx:70]: I'm fine with that, no real attachment to where it's at right now.

File: components/dashboard/src/user-settings/Preferences.tsx
Comment (gtsiolis) [reply components/dashboard/src/user-settings/Preferences.tsx:90]: Ah, I see. Yeah, let's make this change if it's trivial—otherwise we could keep it like this, as I wouldn't be surprized if the UX or interaction to update the dotfiles or timeouts moves away from these inline forms, see relevant comments:
- https://github.com/gitpod-io/gitpod/issues/7385#issuecomment-1278860196
- https://github.com/gitpod-io/gitpod/issues/7385#issuecomment-1372436421

**Pull request merged**
